fem/hyperbolic.hpp

```diff
@@ -0,0 +1,472 @@
+    * @param[in] Tr face information
+    * @param[out] flux numerical flux (num_equations)
+    */
+   virtual double Eval(const Vector &state1, const Vector &state2,
+                       const Vector &nor, FaceElementTransformations &Tr,
+                       Vector &flux) const = 0;
+
+   virtual ~RiemannSolver() = default;
+
+   /// @brief Get flux function F
+   /// @return constant reference to the flux function.
+   const FluxFunction &GetFluxFunction() const { return fluxFunction; }
+
+protected:
+   const FluxFunction &fluxFunction;
+};
+
+/**
+ * @brief Abstract hyperbolic form integrator, (F(u, x), ∇v) and (F̂(u±, x, n))
+ *
+ */
+class HyperbolicFormIntegrator : public NonlinearFormIntegrator
+{
+private:
+   // The maximum characterstic speed, updated during element/face vector assembly
+   double max_char_speed;
+   const RiemannSolver &rsolver;   // Numerical flux that maps F(u±,x) to hat(F)
+   const FluxFunction &fluxFunction;
+   const int IntOrderOffset; // integration order offset, see GetRule()
+#ifndef MFEM_THREAD_SAFE
+   // Local storages for element integration
+   Vector shape;              // shape function value at an integration point
+   Vector state;              // state value at an integration point
+   DenseMatrix flux;          // flux value at an integration point
+   DenseMatrix dshape;  // derivative of shape function at an integration point
+
+   Vector shape1;  // shape function value at an integration point - first elem
+   Vector shape2;  // shape function value at an integration point - second elem
+   Vector state1;  // state value at an integration point - first elem
+   Vector state2;  // state value at an integration point - second elem
+   Vector nor;     // normal vector, @see CalcOrtho
+   Vector fluxN;   // hat(F)(u,x)
+#endif
+
+public:
+   const int num_equations;  // the number of equations
+   /**
+    * @brief Construct a new Hyperbolic Form Integrator object
+    *
+    * @param[in] rsolver numerical flux
+    * @param[in] IntOrderOffset integration order offset, see GetRule()
+    */
+   HyperbolicFormIntegrator(
+      const RiemannSolver &rsolver,
+      const int IntOrderOffset=0);
+
+   /**
+    * @brief Get the element integration rule based on IntOrderOffset, see
+    * AssembleElementVector(). Used only when ir is not provided
+    *
+    * @param[in] el given finite element
+    * @param[in] Tr element transformation for Jacobian order
+    * @param[in] IntOrderOffset_ integration order offset
+    * @return const IntegrationRule& with order 2*p + Tr.OrderJ() +
+    *    IntOrderOffset
+    */
+   static const IntegrationRule &GetRule(const FiniteElement &el,
+                                         const ElementTransformation &Tr,
+                                         const int IntOrderOffset_)
+   {
+      const int order = 2 * el.GetOrder() + Tr.OrderJ() + IntOrderOffset_;
+      return IntRules.Get(Tr.GetGeometryType(), order);
```

Comment:
We should discuss what's a good default choice for the quadrature rule. For example, I'm not convinced that we need any dependence on the mesh degree which is what `Tr.OrderJ()` brings.

Reply:
$(F(u), ∇ϕ)_T = (F(u), Jᵀ∇̂ϕ̂*w)_T̂$
If we assume $F(u)$ is a polynomial of order `el.order() + IntOrderOffset_`, then the total order would be
`(el.GetOrder() + IntOrderOffset_) + (Tr.OrderJ()) + (el.GetOrder() - ?) + (Tr.OrderW())`.
This will produce exact integration. Numerically, exact integration is often not required to expect optimal convergence behavior.

I'm curious whether we can just remove the dependency on `Tr` when curved elements are used. I can see that there are several `BilinearFormIntegrator` that uses `Tr` to obtain the integration order,  e.g.,`MassIntegrator`, `ConvectionIntegrator`, `MixedCurlIntegrator`, `VectorFEMassIntegraotr`, `VectorDivergenceIntegrator`, `DGTraceIntegrator`.
I have never used curved elements before, so I would like to hear some opinions from others who have used curved elements.

Reply:
For some more context, the term $`\left(F(u),\nabla \phi\right)_T`$ transformed to the reference element becomes
```math
\left(F(u),\nabla \phi\right)_T =
   \left(F(\hat{u}), \det(J) J^{-T} \hat{\nabla} \hat{\phi}\right)_{\hat{T}} =
   \left(F(\hat{u}), \mathrm{adj}(J)^T \hat{\nabla} \hat{\phi}\right)_{\hat{T}}
```
and the method `ElementTransformation::OrderGrad(const FiniteElement *)` returns the order (polynomial degree) of the entries of the term $g = \mathrm{adj}(J)^T \hat{\nabla} \hat{\phi}$ for the given `FiniteElement` argument describing $\hat\phi$. However, the degree of the entries of $g$ grows fast with the degree of the mesh and the space dimension:
* $`g\in P_p`$ with $p=(k-1)(d-1)+(l-1)$, for $(P_k)^d$ mesh element transformation and $\hat\phi\in P_l$ test function,
* $`g\in Q_p`$ with $p=k(d-1)+(l-1)$, for $(Q_k)^d$ mesh element transformation and $\hat\phi\in Q_l$ test function.

For this reason, exact integration (assuming $F(\hat u)$ is also polynomial) is not practical.

Reply:
Thank you for detailed explanation. I agree that the exact integration is not practical. I tested various integration order on a quadratic mesh with cubic FE, and confirmed that the accuracy difference was negligible with/without `OrderGrad` or other transform related orders.
<img width="371" alt="image" src="https://github.com/mfem/mfem/assets/17425440/66f27f0f-8e74-463a-b4ba-605a9fdf70fe">

The updated version now only uses polynomial order and user-defined `IntOrderOffset` to determine the integration order, `2*el.GetOrder() + IntOrderOffset` for elements, and `2*std::max(el1.GetOrder(), el2.GetOrder()) + IntOrderOffset` for interfaces.